core/string/src/builder.rs

```diff
@@ -0,0 +1,805 @@
+use crate::{
+    alloc_overflow, tagged::Tagged, JsStr, JsStrVariant, JsString, RawJsString, RefCount,
+    TaggedLen, DATA_OFFSET,
+};
+
+use std::{
+    alloc::{alloc, dealloc, realloc, Layout},
+    cell::Cell,
+    marker::PhantomData,
+    mem::ManuallyDrop,
+    ops::AddAssign,
+    ptr::{self, addr_of_mut, NonNull},
+    str::{self},
+};
+
+/// A mutable builder to create instance of `JsString`.
+///
+#[derive(Debug)]
+pub struct JsStringBuilder<D> {
```

Comment:
Can we add the restriction that `D: Copy`? Just to be safe, because we still want to nudge users to the correct types (`u8` or `u16`)